### DISCUSSION TITLE
fmiSimulateCS did not work with fmiSetReal?

### DISCUSSION THREAD

**Discussion by Amuri44** [ThummeTo/FMI.jl#71] [Q&A]
I have a simple fmu file which is addition module (for testing) with two inputs `var1,var2` and single output `y`. I have three issues:

1- When I run it, the execution time is 0.068864 second for the `fmiSimulate(myFMU, 0.0, 0.0; recordValues="y")` function. It is kind of slow. Any reason why and how to speed it up?
2- I tried to change the values of inputs and simulate, as below. However, the `fmiSetReal` worked but`fmiSimulate` did not take the new values and returned zero values (default). Any reason for that and how to solve the problem?
```julia
# Model built with fmuType = "me_cs" (both model exchange and co-simulation)
@time begin 
# change var1 and var2
success, simData = fmiSimulateCS(myFMU, 0.0, 0.0; recordValues=["y"])
end
```
3- When using `fmiSimulateME()` Instead of `fmiSimulate()`, it gives error. Any reason for that?
```julia
success, simData = fmiSimulateME(myFMU, 0.0, 0.0; setup=true)
ERROR: LoadError: BoundsError: attempt to access 0-element Vector{Float64} at index [1]
Stacktrace:
```

**Comment by ThummeTo** [comment]
Dear @Amuri44 ,

as far as I understand you correct, this is not really an issue. 

(2,3) FMI is for simulating dynamic models, so a simulation from `t_start=0.0s` to `t_stop=0.0s` may get never evaluated (dependent on the exporting tool). Use the keyword `setup=false` for `fmiSimulate` if you want to do your setup by your own (have a look at the examples-folder, better documentation coming soon). FMI.jl does not simulate, because there is no delta time. Try to simulate one second and have a look on the results. Nevertheless, I think `setup=false` will do the trick and you don't even need to call `fmiSimulate` in your special case.

(1) Further, you are right, an addition-FMU is slow compared to a pure-C-addition-statement, but this should be expected, because there is much much more happening in the background, than just the simple addition process. 

If you really just want to add two numbers, FMI is not the right way to do. If you have larger systems and benchmark the simulation time compared to the simulation time in pure C, the timing will be much better, because a FMU is basically compiled C-Code.

Best regards!

EDIT: I moved this to discussions.

**Reply by Amuri44** [reply]
Dear @ThummeTo ,

Thank you for your reply!
The addition is just a toy here not my real model. But I found by this simple example that the execution time of 0.068864 second kind of slow if I care about the performance in my code.
1- Excuse me, I did catch the point here that when using fmiSimulateME() Instead of fmiSimulate(), it gives error?
```julia
success, simData = fmiSimulateME(myFMU, setup=false)
ERROR: LoadError: BoundsError: attempt to access 0-element Vector{Float64} at index [1]
Stacktrace
```
2- So far, I can only use `fmiSimulateCS` to run `myFMU` over a period of time. However, this `myFMU` has inputs and outputs and contains integrators and other nonlinear elements. I have a time loop in the julia code, in which I modify the inputs of `myFMU` based on its outputs at the previous time step. How can I let `myFMU` runs over a simulation time but at each time step I exchange data with my julia code?

```
3- Can I transfer multiple variables through `fmiSetReal` or it is fixed to single one?

**Comment by ThummeTo** [comment]
1) The correct call is `solution = fmiSimulateME(myFMU, setup=false)` because an ODE is solved in this case. If you want to record non-state values, the call is `solution, savevalues = fmiSimulateME(myFMU, setup=false; recordValues=[...])`.

2) if you want to build a custom simulation loop, just have a look at the implementation of `fmi2SimulateCS` in `FMI2_sim.jl`, this is a very good and easy to understand starting point (it is not common to use `fmiEnterInitializationMode` or `fmiExitInitializationMode` inside of the loop, this may cause errors by the FMU itself). In general, the FMI specification is very good literature if you are interessted in what FMI is capable of: [link](https://fmi-standard.org/)

3) you can transfer multiple variables, just use arrays instead of scalars :-)

Best regards!

**Reply by Amuri44** [reply]
Dear @ThummeTo 

> 2\. if you want to build a custom simulation loop, just have a look at the implementation of `fmi2SimulateCS` in `FMI2_sim.jl`, this is a very good and easy to understand starting point.

1- I will have a look. However, could you please give me a simple example or simply convert this simple example: [link](https://github.com/ThummeTo/FMI.jl/discussions/71#:~:text=a%20correct%20way%3F-,s_time%20%3D%200%3B%0Ae_time%20%3D%201%3B%0Astep%3D0%2C1%3B%0AK%3D%201%3B%20%23%20intinal%20value,fmiDoStep(myFMU%2C%20t%2C%20step)%0A%23%20change%20K%20based%20on%20y%0Aend,-3%2D%20Can%20I)

2- If a custom simulation loop can be build  by `fmi2SimulateCS` what is the importance use of `fmiDoStep`?

> (it is not common to use `fmiEnterInitializationMode` or `fmiExitInitializationMode` inside of the loop, this may cause errors by the FMU itself).

Yes, when I implemented it with time-step less than 0.021, julia does not give results but runs in an endless loop.

**Comment by ThummeTo** [comment]
Dear @Amuri44 ,

`fmiSimulateCS` performs a CS-simulation of the FMU (this command was written by us).
`fmiDoStep` does one time step with the FMU (this command wraps directly a call of the FMI-specification, see the spec for more information).

Some users might want to do very customized stuff during simulation, so you can build a custom loop using `fmiDoStep` or just simulate the entrie FMU using `fmiSimulateCS`.

The function definition of `fmiSimulateCS` is basically a simple example on how to simulate a CS-FMU using `fmiDoStep`. Also the example under `example/CS_simulate.jl` gives an example on how to use `fmiSimulateCS`. 

If you only wnat to change a FMU input, this is already supported by using:

`success, solution = fmiSimulateCS(fmuStruct, t_start, t_stop; saveat=t_start:dt:t_stop, recordValues=["mass.s", "mass.v"], inputValues=["extForce"], inputFunction=extForce)`

where `["extForce"]` is an array of input names (or variable identifiers) and `extForce` is a custom function that returns the input values for a given time `t`, like for example:

```
function extForce(t)
    [sin(t)]
end  
```

If you have a high-frequency input you may want to decrease the `dt` so the FMU is forced to react to the input changes. 

Best regards!

**Reply by Amuri44** [reply]
Dear @ThummeTo Thank you very much for your details
> `fmiDoStep` does one time step with the FMU (this command wraps directly a call of the FMI-specification, see the spec for more information).
I just want to double check, if the implementation of my MWE of time-loop using `fmiDoStep` is correct based on what I have learnt from the implementation of `fmiSimulateCS` as below:
If myFMU (contains integrators and other nonlinear elements) has input `var1` and output `y`. At each time-step I need to modify the input of myFMU based on its output from the previous time step.

**Reply by ThummeTo** [reply]
Looks fine on the first look!

Note, that this can be done also with:
```
function myK(t)
   return K * fmiGetReal(myFMU, "y") # change K dependent on y (and t if you want)
end
success, solution = fmiSimulateCS(myFMU, s_time, e_time; saveat=s_time:step:e_time, recordValues=["y"], inputValues=["var1"], inputFunction=myK)
```

Best regards!

**Reply by Amuri44** [reply]
Thank you very much!

I have a model designed in OpenModelica contains integrators and transfer functions and first orders. When I run this model in OMEdit, it works fine with some warnings: 
```modelica
Compilation process finished successfully

The initialization finished successfully without homotopy method.
### STATISTICS ###
The simulation finished successfully.

[2] 13:55:36 Translation Warning
Assuming fixed start value for the following 1 variables:
        dd.firstOrder.y:VARIABLE(flow=false start = dd.firstOrder.y_start fixed = true )  "Connector of Real output signal" type: Real
```
However, when I run it from Julia via FMI (in similar structure of [link](https://github.com/ThummeTo/FMI.jl/discussions/71#:~:text=s_time%20%3D%200%3B%0Ae_time%20%3D%201%3B%0Astep%3D0.1,change%20%60K%60%20based%20on%20%60y%60%0Aend); time step dt=0.01), it is stuck running at a specific time step with the following repeating message:
```julia
assert    | debug   | division leads to inf or nan at time 0.056, (a=7.2e+302) / (b=1e-15), where divisor b is: dd.firstOrder.T   | division leads to inf or nan at time 0.056, (a=7.2e+302) / (b=1e-15), where divisor b is: firstOrder.T
assert    | debug   | division leads to inf or nan at time 0.056, (a=7.2e+302) / (b=1e-15), where divisor b is: dd.firstOrder.T
....
assert    | debug   | division leads to inf or nan at time 0.056, (a=7.2e+302) / (b=1e-15), where divisor b is: dd.firstOrder.T
```
It seems that something related to block `firstOrder` is being fixed automatically by OMEdit but not when running FMI in julia, correct? the problem is not issued when (dt >=0.03 s) Any idea how to solve this issue?

**Reply by ThummeTo** [reply]
This is very well explained by the error message itself: Division by zero, as you can see in `a/b` the variabel `a` becomes very big (`7.2e302`) and variable `b` becomes very small `1e-15`, so the result is close to undefined (division by zero) or close to infinity (because Float64 is limited to I think around `1e308`). Try to find out what `a` and `b` are.

Is this message generated by Julia (so division happens in Julia) or the FMU (so divison happens in the FMU)? Please check what you are setting. I fthis is a very sensitive closed-loop simulation, it is very important to set and get values at the right time, otherwise the simulation can become instable. Without knowing the entire system and code it is hard to debug ...

**Reply by Amuri44** [reply]
Thank you!
